### ISSUE TITLE
An otherwise normal terraform error/failure causes the assume provider to emit a stacktrace/crash.

### ISSUE THREAD

**Issue by craigmiskell-gitlab** [apparentlymart/terraform-provider-assume#1]
If the `assume` provider is included, any error from terraform will result in a stack trace from the assume provider;

```
Stack trace from the terraform-provider-assume_v0.1.0 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x89906d]
goroutine 8 [running]:
go.rpcplugin.org/rpcplugin/internal/gopluginshim.(*controllerServer).Shutdown(0xe9c7c0?, {0x952f80?, 0xc000170998?}, 0x49ab85?)
	go.rpcplugin.org/rpcplugin@v0.3.0/internal/gopluginshim/gopluginshim.go:25 +0xd
go.rpcplugin.org/rpcplugin/internal/gopluginshim._GRPCController_Shutdown_Handler({0x952f80, 0xc000066060}, {0xad9f78, 0xc000616150}, 0xc000122400, 0x0)
	go.rpcplugin.org/rpcplugin@v0.3.0/internal/gopluginshim/gp_controller.pb.go:134 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00020ee00, {0xad9f78, 0xc0006160c0}, {0xadd9a0, 0xc000124000}, 0xc000614000, 0xc0001336e0, 0xe9c7a0, 0x0)
	google.golang.org/grpc@v1.62.0/server.go:1383 +0xdd1
google.golang.org/grpc.(*Server).handleStream(0xc00020ee00, {0xadd9a0, 0xc000124000}, 0xc000614000)
	google.golang.org/grpc@v1.62.0/server.go:1794 +0xe87
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.62.0/server.go:1027 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 57
	google.golang.org/grpc@v1.62.0/server.go:1038 +0x125
Error: The terraform-provider-assume_v0.1.0 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
2024/10/31 01:23:00 terraform returned an exit code of 1
```

It can be reproduced with something as simple as:

```
terraform {
  required_providers {
    assume = {
      source  = "apparentlymart/assume"
      version = ">= 0.1.0"
    }
  }
}

resource "local_file" "configs" {
  filename = "/non-writable-dir/crash.txt"
  content = "crash"
}
```

where the local_file cannot create that directory (assuming you're not running as root).

**Comment by apparentlymart** [comment]
Thanks for reporting this, @craigmiskell-gitlab!

This seems to be a bug in the upstream library `go.rpcplugin.org/rpcplugin`, but fortunately I _also_ maintain that library so I'll see what I can do to fix it soon and then publish a new release of this provider.

It seems like a relatively straightforward bug: [this statement to instantiate the service that handles Terraform's "shutdown" request](https://github.com/rpcplugin/go/blob/8d0547d88b75526c3d460e26921649f1cdbcfd16/internal/gopluginshim/gopluginshim.go#L16) is not passing the `close` callback into the `controllerServer` instance, and so when Terraform instructs the plugin to shutdown it tries to call a nil function pointer. :man_facepalming: 

(This bug presumably also affects the `apparentlymart/hcl` plugin, which uses the same upstream library.)

**Comment by craigmiskell-gitlab** [comment]
> fortunately I also maintain that library

@apparentlymart Convenient! :)

Thanks for the quick response.

**Comment by apparentlymart** [comment]
Hi again @craigmiskell-gitlab!

I've just published v0.1.1 of this provider which should address this problem. The Terraform public registry has the latest version already, and the OpenTofu registry should follow at some point once it resynchronizes with the latest releases.

Thanks for reporting this and sorry for the delay in fixing it.

**Comment by craigmiskell-gitlab** [comment]
> Thanks for reporting this and sorry for the delay in fixing it.

@apparentlymart No worries at all, I appreciate you sharing your code in the first place, and taking the time to update it.
